IND-4379 ✅ feat(TfWorkspaceStateUtility): Add workspace state handling utility functions to convert the workspace state to stack state

sujaysamanta opened the pull request:
## Background

Add the following functions to the https://github.com/hashicorp/terraform-migrate-utility library.
```go
// IsFullyModular checks if the provided workspace state data is fully modular.
// Acceptance Criteria:
// - Input: A byte slice representing the workspace state data.
// - Output: A boolean indicating whether the workspace state is fully modular, and an error if any issue occurs during processing.
// - The function should return true if all modules are properly defined and no resources are outside of the modules.
// - The function should return false if any resources are found outside of modules.
func (tf *tfWorkspaceStateUtility) IsFullyModular(managedResourcesJson []byte) bool {
	panic("implement me")
}

// WorkspaceToStackAddressMap converts the workspace state data to a map of workspace addresses to stack addresses.
// Acceptance Criteria:
// - Input: A byte slice representing the workspace state data and a string representing the absolute path to the stack source bundle.
// - Output: A map where keys are workspace addresses and values are corresponding stack addresses.
// - The function should correctly map each workspace address to its respective stack address based on the provided state data.
// - If the input data is invalid, the function should handle errors gracefully and return an empty map.
func (tf *tfWorkspaceStateUtility) WorkspaceToStackAddressMap(workspaceStateData []byte, stackSourceBundleAbsPath string) (WorkspaceStateAddressMap, error){
	panic("implement me")
}
```

https://hashicorp.atlassian.net/browse/IND-4379


## PCI review checklist

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

- [x] I have documented a clear reason for, and description of, the change I am making.

- [ ] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.

- [ ] If applicable, I've documented the impact of any changes to security controls.

  Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

Commit:
port rpcapi client

Commit:
add dependabot and PR template

Commit:
lint and test

Commit:
remove workflows

Commit:
Stack State Proto

Commit:
refactor

Commit:
readme update

sujaysamanta committed:
feat: Initial commit

sujaysamanta committed:
Merge branch 'main' into mod-utility

sujaysamanta committed:
:sparkles: feat(TfWorkspaceStateUtility): Added an interface for operations related to Terraform workspace state.

sujaysamanta committed:
:white_check_mark: feat(TfWorkspaceStateUtility): Implemented `GetManagedResources` and `IsFullyModular` methods

sujaysamanta committed:
feat(TfWorkspaceStateUtility): Added WorkspaceToStackAddressMap to utility functions and HCL parsing to TfWorkspaceStateUtility interface

sujaysamanta committed:
Merge branch 'main' into IND-4379-Add-Workspace-State-Handling-Utility-Functions

sujaysamanta committed:
fix: updated go.mod and go.sum

a-anurag27 reviewed (changes requested):
Some items noted post e2e testing

sujaysamanta committed:
:white_check_mark: feat(TfWorkspaceStateUtility): Introduced `ListAllResourcesFromWorkspaceState` and overhauled `WorkspaceToStackAddressMap` for modular state handling

- Refactored `WorkspaceToStackAddressMap` implementation to improve modularization checks and resource-to-stack mapping logic.
- Added new method `ListAllResourcesFromWorkspaceState` to retrieve all resources from the workspace state using Terraform CLI.
- Simplified `IsFullyModular` logic, ensuring better modular resource validation.
- Enhanced examples in `README.md` to demonstrate updated functionalities.

sujaysamanta committed:
Merge branch 'main' into IND-4379-Add-Workspace-State-Handling-Utility-Functions
